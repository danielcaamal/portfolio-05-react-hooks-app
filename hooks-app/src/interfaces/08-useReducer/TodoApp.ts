

export interface TodoInterface {
    id: number;
    description: string;
    done: boolean;
};

export interface TodoListInterface {
    todos: TodoInterface[];
    onDeleteTodo: (id:number) => void;
    onToggleTodo: (id:number) => void;
}

export interface TodoItemInterface {
    todo: TodoInterface;
    onDeleteTodo: (id:number) => void;
    onToggleTodo: (id:number) => void;
}
