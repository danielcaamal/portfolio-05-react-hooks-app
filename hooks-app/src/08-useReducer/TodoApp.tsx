import { TodoList, TodoAdd } from './';
import { useTodo } from '../hooks/useTodo';

export const TodoApp = () => {
    const { todos, todosCount, pendingTodosCount,
        handleDeleteTodo, handleToggleTodo, handleNewTodo } = useTodo();
    
    return (
        <>
            <h1>TodoApp: {todosCount}, <small>pendings: </small>{pendingTodosCount}</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <h4>Todos</h4>
                    <hr />
                    <TodoList 
                        todos={todos} 
                        onDeleteTodo={handleDeleteTodo} 
                        onToggleTodo={handleToggleTodo}
                        />
                </div>
                <div className="col-5">
                    <h4>Add Todo</h4>
                    <hr />
                    <TodoAdd 
                        onNewTodo={handleNewTodo} />
                </div>
            </div>
        </>
    ); 
}