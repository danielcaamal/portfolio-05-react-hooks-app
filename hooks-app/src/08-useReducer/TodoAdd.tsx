import { FormEvent } from "react";
import { useForm } from "../hooks/useForm";
import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

export const TodoAdd = ({ onNewTodo }: { onNewTodo: (todo:TodoInterface) => void }) => {

    const initialTodo = {
        id: 0,
        description: '',
        done: false
    }
    const { formState, onInputChange, onResetForm } = useForm(initialTodo)
    const { description } = formState;

    const onSubmit = (e: FormEvent) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;

        const newTodo = {
            ...initialTodo,
            id: new Date().getTime(),
            description,
        };

        onNewTodo(newTodo);
        onResetForm();
    }

    return (
        <>
            <form
                id="todo-add-form" 
                onSubmit={(e) => onSubmit(e)}>
                <input
                    id="todo-add-input"
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Learn..."
                    autoComplete="off"
                    onChange={onInputChange}
                    value={description}
                />
                <button
                    id="todo-add-button"
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                    >Add
                </button>
            </form>
        </>
    );
}