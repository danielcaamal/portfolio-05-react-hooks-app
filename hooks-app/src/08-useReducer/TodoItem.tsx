import { TodoItemInterface } from "../interfaces/08-useReducer/TodoApp";

export const TodoItem = (
    { 
        todo, 
        onDeleteTodo,
        onToggleTodo 
    } : TodoItemInterface) => {

    return (
        <>
            <span 
                id="todo-item-description"
                className={`align-self-center ${todo.done && 'text-decoration-line-through'}`}
                onDoubleClick={() => onToggleTodo(todo.id)}
                >{todo.description}</span>
            <button
                id="todo-item-delete-button"
                className="btn btn-danger"
                type="button"
                onClick={() => onDeleteTodo(todo.id)}
                >Delete
            </button>
        </>
    );
}