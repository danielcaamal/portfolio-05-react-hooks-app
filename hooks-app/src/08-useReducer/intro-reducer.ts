import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

const initialState : TodoInterface[] = [
    { id: 1, description: 'Buying a car', done: false },
    { id: 2, description: 'Buying a house', done: false },
]

const todoReducer = (state: TodoInterface[] = initialState, action?: {type?: string, payload: TodoInterface}) => {
    switch (action?.type) {
        case 'add':
            return [...state, action.payload]
        default:
            break;
    }

    return state;
}

let todos = todoReducer(initialState);

const newTodo = {
    id: 3,
    todo: 'Buying a laptop',
    done: false
}

const addTodoAction: any = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, addTodoAction)