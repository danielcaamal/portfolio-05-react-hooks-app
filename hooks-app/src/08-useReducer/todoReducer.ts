import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

export const todoReducer = (state: TodoInterface[] = [], action: { type:string, payload: TodoInterface }): TodoInterface[] => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter(todo => todo.id !== action.payload.id);
        case 'toggle':
            return state.map(
                (todo) => {
                    if (todo.id === action.payload.id) {
                        return {
                            ...todo,
                            done: !todo.done
                        }
                    }
                    return todo;
                }
            )
        default:
            return state;
    }
}