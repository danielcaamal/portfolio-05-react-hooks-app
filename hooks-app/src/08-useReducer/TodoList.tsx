import { TodoListInterface } from '../interfaces/08-useReducer/TodoApp';
import { TodoItem } from './TodoItem';

export const TodoList = (
    {
        todos = [], 
        onDeleteTodo,
        onToggleTodo
    } : TodoListInterface) => {

    return (
        <>
            <ul className="list-group">
                {todos.map((todo, i) => {
                    return (
                        <li 
                            className="list-group-item d-flex justify-content-between" 
                            key={ todo.id }>
                            <TodoItem 
                                todo={todo} 
                                onDeleteTodo={onDeleteTodo}
                                onToggleTodo={onToggleTodo}
                                />
                        </li>
                    )
                })}
            </ul>
        </>
    );
}