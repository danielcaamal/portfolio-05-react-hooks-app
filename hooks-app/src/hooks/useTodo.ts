import { useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";
import { TodoInterface } from "../interfaces/08-useReducer/TodoApp";

const initialState : TodoInterface[] = []

const init = () : TodoInterface[] => {
    return JSON.parse(localStorage.getItem('todos') ?? '[]');
}


export const useTodo = () => {
    const [todos, dispatchTodos] = useReducer(todoReducer, initialState, init)

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleNewTodo = (newTodo: TodoInterface) => {
        console.log(newTodo);
        const action = {
            type: 'add',
            payload: newTodo
        };
        dispatchTodos(action);
    }

    const handleDeleteTodo = (id: number) => {
        const todo = todos.find(todo => todo.id === id);
        if (!todo) return;
        const action = {
            type: 'delete',
            payload: todo
        };
        dispatchTodos(action);
    }

    const handleToggleTodo = (id: number) => {
        const todo = todos.find(todo => todo.id === id);
        if (!todo) return;
        const action = {
            type: 'toggle',
            payload: todo
        };
        dispatchTodos(action);
    }

    const todosCount = todos.length;
    const pendingTodosCount = todos.filter((todo) => !todo.done).length;

    return {
        todos,
        todosCount,
        pendingTodosCount,
        handleNewTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
}