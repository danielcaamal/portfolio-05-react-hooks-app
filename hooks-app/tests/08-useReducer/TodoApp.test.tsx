import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { TodoApp } from '../../src/08-useReducer/TodoApp';
import * as UseTodo from '../../src/hooks/useTodo';

describe('TodoApp', () => {

    const spy = jest.spyOn(UseTodo, 'useTodo');

    spy.mockReturnValue({
        todos: [
            { id: 1, description: 'Learn React', done: false },
            { id: 2, description: 'Learn Mongo', done: false },
        ],
        todosCount: 2,
        pendingTodosCount: 2,
        handleDeleteTodo: jest.fn(),
        handleToggleTodo: jest.fn(),
        handleNewTodo: jest.fn(),
    });


    test ('should match with snapshot', () => {
        const { container } = render(<TodoApp />);
        expect(container).toMatchSnapshot();
    });

    test ('should show 2 todos', () => {
        const { container } = render(<TodoApp />);
        expect(screen.getByText('Learn React')).toBeTruthy();
        expect(screen.getByText('Learn Mongo')).toBeTruthy();
    });

    test ('should call add a todo', () => {
        const { container } = render(<TodoApp />);
        const input = container.querySelector('#todo-add-input') as HTMLInputElement;
        const form = container.querySelector('#todo-add-form') as HTMLFormElement;
        fireEvent.change(input, { target: { value: 'Learn Redux' } });
        fireEvent.submit(form);
        expect(spy).toHaveBeenCalled();
    });
});
