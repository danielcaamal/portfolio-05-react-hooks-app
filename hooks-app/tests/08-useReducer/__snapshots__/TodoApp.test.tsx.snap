// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp should match with snapshot 1`] = `
<div>
  <h1>
    TodoApp: 
    2
    , 
    <small>
      pendings: 
    </small>
    2
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <h4>
        Todos
      </h4>
      <hr />
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center false"
            id="todo-item-description"
          >
            Learn React
          </span>
          <button
            class="btn btn-danger"
            id="todo-item-delete-button"
            type="button"
          >
            Delete
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center false"
            id="todo-item-description"
          >
            Learn Mongo
          </span>
          <button
            class="btn btn-danger"
            id="todo-item-delete-button"
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add Todo
      </h4>
      <hr />
      <form
        id="todo-add-form"
      >
        <input
          autocomplete="off"
          class="form-control"
          id="todo-add-input"
          name="description"
          placeholder="Learn..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1 btn-block"
          id="todo-add-button"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
